### PR TITLE
executor: run idxlookup workers in a pool

### PR DESCRIPTION
**Pull request by zyguan**
### What problem does this PR solve?

Issue Number: ref #56649

Problem Summary: IndexLookupExecutor always starts 1+indexLookupConcurrency goroutines which is expensive for common queries.

### What changed and how does it work?

Use a goroutine pool for workers and spawn new goroutine on demand.

Let prepare sysbench dataset with `--secondary` and run point select workload.

Previously, tidb spawns 6 goroutines for each query and only two of them are used (non-concurrently).

image

With this PR, only 1 goroutine are created for both index task and table task.

Here are benchmark results.

|     workload     | threads |   pool   |   base   |  diff  |
| ---------------- | ------- | -------: | -------: | -----: |
| idxlookup_1      | 15      | 16560.51 | 16320.65 |  1.47% |
| idxlookup_1      | 30      | 26729.51 | 26146.49 |  2.23% |
| idxlookup_1      | 60      | 29680.83 | 28253.65 |  5.05% |

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by crazycs520 — APPROVED** [review]

File: pkg/executor/utils.go [hunk 1]
```diff
@@ -119,3 +121,70 @@ func encodePassword(u *ast.UserSpec, authPlugin *extension.AuthPlugin) (string,
...
+}
+
+type gotask struct {
+	f    func()
+	next *gotask
+}
+
+type gopool struct {
+	lock sync.Mutex
+	head *gotask
```
Comment (cfzjywxk) [inline pkg/executor/utils.go:136]: How about using a lock free queue to store the pending task pointers, so the `Mutex` could be avoided?

File: pkg/executor/utils.go [hunk 2]
```diff
@@ -119,3 +121,70 @@ func encodePassword(u *ast.UserSpec, authPlugin *extension.AuthPlugin) (string,
...
+var taskPool = sync.Pool{
+	New: func() any { return &gotask{} },
+}
+
+type gotask struct {
+	f    func()
+	next *gotask
+}
+
+type gopool struct {
```
Comment (cfzjywxk) [inline pkg/executor/utils.go:134]: There is some `GoroutinePool` like structures in other packages, how about renaming it to `IndexLookupWorkerPool` or `CopWorkerPool` it we do not want it to be a generic used type?

File: pkg/executor/utils.go (hunk 1 above)
Comment (zyguan) [reply pkg/executor/utils.go:136]: Here gopool is mainly for concurrency control, each IndexLookupExecutor has its own pool, that is, there are only about indexLookupConcurrency goroutines access the pool at most 2xNumOfTableTasks times. I think it's acceptable, no need to introduce additional complexity.

File: pkg/executor/utils.go (hunk 2 above)
Comment (zyguan) [reply pkg/executor/utils.go:134]: Actually I prefer to keep it simple and private currently. [gp](https://github.com/tiancaiamao/gp) and [pool](https://github.com/pingcap/tidb/tree/master/pkg/resourcemanager/pool) have already been used in tidb, but neither of them meet my requirement. So I just implement a tiny one here and maybe we can extract uniform GoroutinePool interface in the future.

**Review by cfzjywxk — APPROVED** [review]

**Pull request merged**